File: impl/java/src/test/java/liff/testutil/Json.java
package liff.testutil;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Minimal test-only JSON reader used for the shared conformance artifacts. */
public final class Json {
    private final String source;
    private int position;

    private Json(String source) {
        this.source = source;
    }

    public static Object read(Path path) throws IOException {
        String source = Files.readString(path, StandardCharsets.UTF_8);
        Json parser = new Json(source);
        Object value = parser.value();
        parser.whitespace();
        if (parser.position != source.length()) {
            throw parser.error("trailing content");
        }
        return value;
    }

    private Object value() {
        whitespace();
        if (position >= source.length()) {
            throw error("expected value");
        }
        return switch (source.charAt(position)) {
            case '{' -> object();
            case '[' -> array();
            case '"' -> string();
            case 't' -> literal("true", Boolean.TRUE);
            case 'f' -> literal("false", Boolean.FALSE);
            case 'n' -> literal("null", null);
            default -> number();
        };
    }

    private Map<String, Object> object() {
        expect('{');
        LinkedHashMap<String, Object> result = new LinkedHashMap<>();
        whitespace();
        if (take('}')) {
            return result;
        }
        while (true) {
            whitespace();
            String key = string();
            whitespace();
            expect(':');
            result.put(key, value());
            whitespace();
            if (take('}')) {
                return result;
            }
            expect(',');
        }
    }

    private List<Object> array() {
        expect('[');
        ArrayList<Object> result = new ArrayList<>();
        whitespace();
        if (take(']')) {
            return result;
        }
        while (true) {
            result.add(value());
            whitespace();
            if (take(']')) {
                return result;
            }
            expect(',');
        }
    }

    private String string() {
        expect('"');
        StringBuilder result = new StringBuilder();
        while (position < source.length()) {
            char character = source.charAt(position++);
            if (character == '"') {
                return result.toString();
            }
            if (character != '\\') {
                result.append(character);
                continue;
            }
            if (position >= source.length()) {
                throw error("unfinished escape");
            }
            char escape = source.charAt(position++);
            switch (escape) {
                case '"', '\\', '/' -> result.append(escape);
                case 'b' -> result.append('\b');
                case 'f' -> result.append('\f');
                case 'n' -> result.append('\n');
                case 'r' -> result.append('\r');
                case 't' -> result.append('\t');
                case 'u' -> result.append((char) hexadecimal(4));
                default -> throw error("invalid escape");
            }
        }
        throw error("unterminated string");
    }

    private int hexadecimal(int digits) {
        if (position + digits > source.length()) {
            throw error("unfinished Unicode escape");
        }
        int result = 0;
        for (int count = 0; count < digits; count++) {
            int digit = Character.digit(source.charAt(position++), 16);
            if (digit < 0) {
                throw error("invalid Unicode escape");
            }
            result = result * 16 + digit;
        }
        return result;
    }

    private Object number() {
        int start = position;
        take('-');
        digits();
        boolean floating = false;
        if (take('.')) {
            floating = true;
            digits();
        }
        if (take('e') || take('E')) {
            floating = true;
            if (!take('+')) {
                take('-');
            }
            digits();
        }
        String text = source.substring(start, position);
        try {
            if (floating) {
                return Double.valueOf(text);
            }
            return Long.valueOf(text);
        } catch (NumberFormatException exception) {
            throw error("invalid number");
        }
    }

    private void digits() {
        int start = position;
        while (position < source.length() && Character.isDigit(source.charAt(position))) {
            position++;
        }
        if (position == start) {
            throw error("expected digit");
        }
    }

    private Object literal(String text, Object value) {
        if (!source.startsWith(text, position)) {
            throw error("invalid literal");
        }
        position += text.length();
        return value;
    }

    private void whitespace() {
        while (position < source.length()) {
            char character = source.charAt(position);
            if (character != ' ' && character != '\n' && character != '\r' && character != '\t') {
                break;
            }
            position++;
        }
    }

    private void expect(char wanted) {
        if (!take(wanted)) {
            throw error("expected '" + wanted + "'");
        }
    }

    private boolean take(char wanted) {
        if (position < source.length() && source.charAt(position) == wanted) {
            position++;
            return true;
        }
        return false;
    }

    private IllegalArgumentException error(String message) {
        return new IllegalArgumentException(message + " at character " + position);
    }
}

File: impl/lua/src/test/lua/json.lua
local M = {}

M.null = setmetatable({}, { __tostring = function() return "null" end })

local Parser = {}
Parser.__index = Parser

local function fail(self, message)
    error(message .. " at byte " .. self.position, 0)
end

function Parser:peek()
    return self.source:sub(self.position, self.position)
end

function Parser:take(character)
    if self:peek() == character then
        self.position = self.position + 1
        return true
    end
    return false
end

function Parser:expect(character)
    if not self:take(character) then
        fail(self, "expected '" .. character .. "'")
    end
end

function Parser:whitespace()
    while self.position <= #self.source do
        local character = self:peek()
        if character ~= " " and character ~= "\n"
            and character ~= "\r" and character ~= "\t"
        then
            return
        end
        self.position = self.position + 1
    end
end

function Parser:hexadecimal()
    local text = self.source:sub(self.position, self.position + 3)
    if #text ~= 4 or not text:match("^[0-9a-fA-F]+$") then
        fail(self, "invalid Unicode escape")
    end
    self.position = self.position + 4
    return tonumber(text, 16)
end

function Parser:string()
    self:expect('"')
    local output = {}
    while self.position <= #self.source do
        local character = self:peek()
        self.position = self.position + 1
        if character == '"' then
            return table.concat(output)
        end
        if character ~= "\\" then
            output[#output + 1] = character
        else
            local escape = self:peek()
            self.position = self.position + 1
            local replacements = {
                ['"'] = '"',
                ["\\"] = "\\",
                ["/"] = "/",
                b = "\b",
                f = "\f",
                n = "\n",
                r = "\r",
                t = "\t",
            }
            if replacements[escape] ~= nil then
                output[#output + 1] = replacements[escape]
            elseif escape == "u" then
                local code_point = self:hexadecimal()
                if code_point >= 0xd800 and code_point <= 0xdbff
                    and self.source:sub(self.position, self.position + 1) == "\\u"
                then
                    self.position = self.position + 2
                    local low = self:hexadecimal()
                    if low < 0xdc00 or low > 0xdfff then
                        fail(self, "invalid low surrogate")
                    end
                    code_point = 0x10000 + (code_point - 0xd800) * 0x400 + (low - 0xdc00)
                end
                output[#output + 1] = utf8.char(code_point)
            else
                fail(self, "invalid string escape")
            end
        end
    end
    fail(self, "unterminated string")
end

function Parser:number()
    local start = self.position
    self:take("-")
    if self:take("0") then
        -- A leading zero is the complete integer part.
    else
        if not self:peek():match("%d") then
            fail(self, "expected digit")
        end
        while self:peek():match("%d") do
            self.position = self.position + 1
        end
    end
    if self:take(".") then
        if not self:peek():match("%d") then
            fail(self, "expected fractional digit")
        end
        while self:peek():match("%d") do
            self.position = self.position + 1
        end
    end
    local exponent = self:peek()
    if exponent == "e" or exponent == "E" then
        self.position = self.position + 1
        if not self:take("+") then
            self:take("-")
        end
        if not self:peek():match("%d") then
            fail(self, "expected exponent digit")
        end
        while self:peek():match("%d") do
            self.position = self.position + 1
        end
    end
    local value = tonumber(self.source:sub(start, self.position - 1))
    if value == nil then
        fail(self, "invalid number")
    end
    return value
end

function Parser:literal(text, value)
    if self.source:sub(self.position, self.position + #text - 1) ~= text then
        fail(self, "invalid literal")
    end
    self.position = self.position + #text
    return value
end

function Parser:array()
    self:expect("[")
    local result = {}
    self:whitespace()
    if self:take("]") then
        return result
    end
    while true do
        result[#result + 1] = self:value()
        self:whitespace()
        if self:take("]") then
            return result
        end
        self:expect(",")
    end
end

function Parser:object()
    self:expect("{")
    local result = { __order = {} }
    self:whitespace()
    if self:take("}") then
        return result
    end
    while true do
        self:whitespace()
        local key = self:string()
        self:whitespace()
        self:expect(":")
        result[key] = self:value()
        result.__order[#result.__order + 1] = key
        self:whitespace()
        if self:take("}") then
            return result
        end
        self:expect(",")
    end
end

function Parser:value()
    self:whitespace()
    local character = self:peek()
    if character == "{" then
        return self:object()
    elseif character == "[" then
        return self:array()
    elseif character == '"' then
        return self:string()
    elseif character == "t" then
        return self:literal("true", true)
    elseif character == "f" then
        return self:literal("false", false)
    elseif character == "n" then
        return self:literal("null", M.null)
    end
    return self:number()
end

function M.decode(source)
    local parser = setmetatable({ source = source, position = 1 }, Parser)
    local result = parser:value()
    parser:whitespace()
    if parser.position <= #source then
        fail(parser, "trailing content")
    end
    return result
end

function M.read(path)
    local file = assert(io.open(path, "rb"))
    local source = assert(file:read("a"))
    file:close()
    return M.decode(source)
end

return M
